-----------

## **1. Setup and Imports**

In this notebook, we will learn how to interact with a locally running Large Language Model (LLM) using Python and Ollama.

Ollama allows us to run language models locally and interact with them through an API. We will use the `ollama` Python library to send prompts to our model and retrieve its responses.

For this project, we will use **Qwen 2.5 3B** as our local language model.

### **Objectives**

* Import the required Python library.
* Establish communication between Python and Ollama.
* Send prompts to the language model and inspect its responses.
* Build reusable functions for interacting with the model.

These fundamentals will prepare us for building a Retrieval-Augmented Generation (RAG) application later.


In [1]:
import ollama

-------------------------

## **2. Connecting to Ollama**

Before sending prompts, we need to ensure that Ollama is accessible from our Python environment.

Ollama runs as a local service and exposes an API that applications can use to interact with installed models.

In our setup, Ollama is running locally, and the `qwen2.5:3b` model is already installed.

### **Objective**

Verify that the Python application can communicate with Ollama and identify the available models.


In [2]:
# Check available models
ollama.list()

ListResponse(models=[Model(model='nomic-embed-text:latest', modified_at=datetime.datetime(2026, 10, 5, 17, 37, 3, 21355, tzinfo=TzInfo(10800)), digest='0a109f422b47e3a30ba2b10eca18548e944e8a23073ee3f3e947efcf3c45e59f', size=274302450, details=ModelDetails(parent_model='', format='gguf', family='nomic-bert', families=['nomic-bert'], parameter_size='137M', quantization_level='F16')), Model(model='qwen2.5:3b', modified_at=datetime.datetime(2026, 10, 4, 13, 38, 51, 42809, tzinfo=TzInfo(10800)), digest='357c53fb659c5076de1d65ccb0b397446227b71a42be9d1603d46168015c9e4b', size=1929912432, details=ModelDetails(parent_model='', format='gguf', family='qwen2', families=['qwen2'], parameter_size='3.1B', quantization_level='Q4_K_M')), Model(model='llama3.2:3b', modified_at=datetime.datetime(2026, 7, 7, 22, 28, 7, 221331, tzinfo=TzInfo(10800)), digest='a80c4f17acd55265feec403c7aef86be0c25983ab279d83f3bcd3abbcb5b8b72', size=2019393189, details=ModelDetails(parent_model='', format='gguf', family='llama

-----------------------

## **3. Sending Our First Prompt**

Now that Ollama is available, we can send our first prompt to the language model.

We will use `ollama.chat()` to provide a model name and a list of messages. Each message contains a role and its content.

In this example:

* `model` specifies which language model to use.
* `messages` contains the conversation input.
* `role="user"` identifies the message as a user request.
* `content` contains the actual prompt.

The model processes the prompt and returns a response object.

### **Objective**

Understand how to send a request to a local LLM and receive its response through Python.


In [ ]:
response = ollama.chat(
	model="qwen2.5:3b",
	messages=[
		{
			"role": "user",
			"content": "What is a Transformer in deep learning?"
		}
	]
)

response

ChatResponse(model='qwen2.5:3b', created_at='2026-10-07T15:26:31.7008326Z', done=True, done_reason='stop', total_duration=73713260300, load_duration=8019373900, prompt_eval_count=37, prompt_eval_duration=1073615000, eval_count=412, eval_duration=64594221000, message=Message(role='assistant', content='A Transformer is a type of neural network architecture developed by Vaswani et al. in their 2017 paper "Attention Is All You Need". It was initially designed as an alternative to recurrent neural networks (RNNs) and convolutional neural networks (CNNs) for sequence processing tasks, such as natural language processing (NLP). Unlike RNNs and CNNs, which use sequential or local operations to process data, Transformers use self-attention mechanisms to weigh the importance of different elements in a sequence simultaneously. This makes Transformers more efficient and effective for tasks like machine translation, text summarization, and natural language understanding.\n\nKey components of a Tran

-----------------

## **4. Understanding the Model Response**

The output of `ollama.chat()` is a response object, not just a plain text string.

This object contains information about the model, the generated message, and generation statistics.

To access the actual answer, we can use:

* `response.message` to access the generated message.
* `response.message.content` to extract the text of the answer.

This distinction is useful because applications usually need the generated text rather than the entire response object.

### **Objective**

Learn how to extract the model's answer from the response object.


In [4]:
# Print only the response
response.message.content

'A Transformer is a type of neural network architecture developed by Vaswani et al. in their 2017 paper "Attention Is All You Need". It was initially designed as an alternative to recurrent neural networks (RNNs) and convolutional neural networks (CNNs) for sequence processing tasks, such as natural language processing (NLP). Unlike RNNs and CNNs, which use sequential or local operations to process data, Transformers use self-attention mechanisms to weigh the importance of different elements in a sequence simultaneously. This makes Transformers more efficient and effective for tasks like machine translation, text summarization, and natural language understanding.\n\nKey components of a Transformer include:\n\n1. **Multi-Head Attention Mechanism**: Allows the model to simultaneously attend to multiple different positions within sequences. This multi-dimensional attention mechanism enables Transformers to handle sequences of variable length and still capture long-range dependencies.\n\n2

In [5]:
# Get the message
response.message

Message(role='assistant', content='A Transformer is a type of neural network architecture developed by Vaswani et al. in their 2017 paper "Attention Is All You Need". It was initially designed as an alternative to recurrent neural networks (RNNs) and convolutional neural networks (CNNs) for sequence processing tasks, such as natural language processing (NLP). Unlike RNNs and CNNs, which use sequential or local operations to process data, Transformers use self-attention mechanisms to weigh the importance of different elements in a sequence simultaneously. This makes Transformers more efficient and effective for tasks like machine translation, text summarization, and natural language understanding.\n\nKey components of a Transformer include:\n\n1. **Multi-Head Attention Mechanism**: Allows the model to simultaneously attend to multiple different positions within sequences. This multi-dimensional attention mechanism enables Transformers to handle sequences of variable length and still cap

------------

## **5. Creating a Reusable Function**

Writing the same API call repeatedly makes code harder to maintain and reuse.

To solve this, we will create a function called `ask_llm()` that accepts a question, sends it to our local language model, and returns the generated answer.

The function will hide the implementation details of the API call behind a simple interface.

### **Example**

Instead of writing the full request every time, we will be able to call:

`ask_llm("What is an embedding?")`

### **Objective**

Encapsulate the Ollama interaction in a reusable function that can be extended as the project grows.


In [ ]:
def ask_llm(question):
	response = ollama.chat(
		model="qwen2.5:3b",
		messages=[
			{
				"role": "user",
				"content": question
			}
		]
	)

	return response.message.content

In [7]:
#ask_llm("What is an embedding?")

'An embedding is a type of mapping that represents information, such as text or images, as vectors of real numbers. These vectors are designed to capture the semantic meaning and context of the information they represent.\n\nIn simpler terms, embeddings are high-dimensional vectors that are meant to capture the essence of data in a way that preserves the important features and relationships between data points. The idea is that these vectors can be used as inputs to machine learning models, enabling them to learn from and make decisions based on complex and nuanced data.\n\nFor example, in natural language processing (NLP), each word can be represented as a high-dimensional vector. These word vectors capture semantic information about the words, such as their meaning, context, and relations (e.g., similar words, synonyms, antonyms). Similarly, images can be represented as vectors, allowing the model to understand the components of an image.\n\nEmbeddings are particularly useful because

In [4]:
ask_llm("What is an embedding? Answer in one sentence.")

'An embedding is a mathematical mapping that translates objects from a source space into a target space, often a high-dimensional vector space, while preserving the semantic relationships between the objects.'

----------------

## **6. Controlling Text Generation**

The length of a generated response can affect both `readability` and `generation time`.

Ollama supports generation options that allow us to control aspects of the model's output.

In this section, we will use `num_predict` to set an upper limit on the number of tokens generated by the model.

We will extend `ask_llm()` with a `max_tokens` parameter so that we can choose a generation limit for each request.

### **Important Note**

`num_predict` limits the number of generated tokens. It does not guarantee a specific response length or generation speed, since performance also depends on the model and available hardware.

### **Objective**

Make our function more flexible by allowing callers to control the maximum number of generated tokens.


In [ ]:
def ask_llm(question, max_tokens=100):
	response = ollama.chat(
		model="qwen2.5:3b",
		messages=[
			{
				"role": "user",
				"content": question
			}
		],
		options={
			"num_predict": max_tokens
		}
	)

	return response.message.content

In [ ]:
ask_llm(
	"Explain self-attention in Transformers.",
	max_tokens=50
)

'Self-attention is a mechanism used in Transformer models to allow them to focus on different parts of the input sequence when making decisions. It is a core component of the attention mechanism used in Transformers, which differs from the traditional recurrent neural networks (RNN'

-------------

## **7. Building a Simple Chat Interface**

So far, we have been manually passing questions to our `ask_llm()` function.

In a real application, users should be able to provide their own questions without modifying the Python code.

In this section, we will build a simple command-line chat interface that repeatedly:

1. Accepts a question from the user.
2. Sends the question to our local LLM.
3. Displays the generated response.
4. Continues until the user chooses to exit.

This will give us a simple interactive interface for our local language model.

### **Objective**

Connect user input to our `ask_llm()` function and create a basic interactive chat experience.


In [ ]:
while True:

	question = input("You: ")

	if question.lower() in ["exit", "quit"]:
		print("Goodbye!")
		break
	
	answer = ask_llm(question, max_tokens = 100)

	print(f"Assistant: {answer}")


Assistant: 你好！我很高兴能在这里帮助你。请问有什么我可以帮助你的吗？如果你有任何问题或需要关于某种技术或主题的信息，请随时告诉我，我会尽力提供帮助。
Assistant: A transformer is a large language model created by Alibaba Cloud. As of my last update, I'm Qwen, an AI assistant designed to provide information and assistance on a wide range of topics. I don't have specific information about the Transformer model itself, but I can explain what a transformer is in general terms.

A transformer, in the context of natural language processing (NLP), is a type of deep learning model architecture developed by Vaswani et al. in 2017.
Goodbye!


----------------------

## **8. Conversation History**

Our current chat interface sends each user question as an independent request.

This means the model does not automatically receive the previous messages in the conversation.

For example, if the user asks:

* "What is a Transformer?"
* "What are its main components?"

The second question may be ambiguous because the model does not know that "its" refers to the Transformer mentioned in the previous message.

To maintain context, we need to store the conversation history and send the previous messages along with each new request.

We will represent the conversation as a list of messages containing different roles:

* `system` — instructions for the assistant.
* `user` — messages written by the user.
* `assistant` — responses generated by the model.

### **Objective**

Understand how conversation history is represented and use it to maintain context across multiple messages.


In [ ]:
def chat_with_history():
	messages = []

	while True :
		question = input("You: ")

		if question.lower() in ["exit", "quit"]:
			print("Goodbye!")
			break

		messages.append({
			"role": "user",
			"content": question
		})

		response = ollama.chat(
			model = "qwen2.5:3b",
			messages = messages,
			options={
				"num_predict": 100
			}
		)

		answer = response.message.content

		messages.append({
			"role": "assistant",
			"content": answer
		})

		print(f"Assistant: {answer}")

In [10]:
chat_with_history()

Assistant: Embedding is a method in machine learning where we represent data, typically in high-dimensional spaces, in a lower-dimensional space, often a 3-dimensional space, by using a learned mapping function. This process is often used in natural language processing (NLP) and other areas where the data has a complex structure.

Here are some key points about embeddings:

1. **Representation**: Embeddings allow us to represent high-dimensional data, such as text or images, in a more manageable, lower-dimensional space
Goodbye!


--------------------------

## **9. Summary and Next Steps**

In this notebook, we learned how to interact with a locally running Large Language Model using Python and Ollama.

We started by connecting Python to Ollama and sending prompts to the `qwen2.5:3b` model. We then learned how to extract the generated response and built reusable functions for interacting with the model.

We also explored how to control text generation using `num_predict` and built a simple interactive chat interface.

Finally, we introduced conversation history by storing previous `user` and `assistant` messages and sending them back to the model with each new request. This allows the model to maintain context throughout a conversation.

### **What We Learned**

* How to communicate with Ollama from Python.
* How to send prompts using `ollama.chat()`.
* How to extract generated text from the response object.
* How to create reusable LLM functions.
* How to control the maximum number of generated tokens.
* How to build a simple command-line chat interface.
* How conversation history provides context across multiple messages.

### **Next Steps**

The next stage of this project is to move from a basic LLM application to a **Retrieval-Augmented Generation (RAG)** system.

Instead of asking the model questions without external knowledge, we will provide it with relevant information retrieved from our own documents.

------------------------

## **RAG Pipeline Architecture**

The following diagram illustrates the main stages of our Retrieval-Augmented Generation (RAG) system.

```text
Documents
	↓
Chunking
	↓
Embeddings
	↓
Vector Database
	↓
Retrieve Relevant Information
	↓
Context + User Question
	↓
Ollama / Qwen 2.5 3B
	↓
Answer
```

### How It Works

1. **Documents:** Load our knowledge sources, such as text files or PDFs.
2. **Chunking:** Split documents into smaller, manageable text segments.
3. **Embeddings:** Convert each text chunk into a numerical vector representing its semantic meaning.
4. **Vector Database:** Store the embeddings and their corresponding text chunks for efficient similarity search.
5. **Retrieve Relevant Information:** Find the chunks most relevant to the user's question.
6. **Context + User Question:** Combine the retrieved information with the user's original question.
7. **Ollama / Qwen 2.5 3B:** Send the question and retrieved context to our locally running language model.
8. **Answer:** Generate a response grounded in the retrieved information.


----------------------------

## **10. Preparing Our Documents**

Before building our RAG pipeline, we need to organize the documents that our application will use as a knowledge source.

In this step, we created a dedicated directory for storing our documents and added a sample text file called `student_handbook.txt`.

### **Project Structure**

```text
ollama-rag-project/
├── data/
│   └── documents/
│       └── student_handbook.txt
├── notebooks/
├── main.py
└── .gitignore
```

### **What We Did**

* Created a `data/` directory to store project data.
* Created a `documents/` subdirectory to organize our source documents.
* Created `student_handbook.txt` as a sample knowledge source.
* Added sample text about course registration, examinations, and graduation requirements.

### **Why Is This Important?**

A RAG application needs access to external information to answer questions using a custom knowledge source.

Organizing our documents separately from our Python code makes the project easier to maintain and extend as we add more documents.

### **Next Step: Document Loading**

Now that our sample document is ready, we will use Python to read its contents into a string.

This will allow us to process the text and prepare it for the next stage: **Chunking**.

```text
student_handbook.txt
		↓
	Python
		↓
	Text String
		↓
	 Chunking
```


In [11]:
# Check your path 
import os

print(os.getcwd())

c:\Users\User\ollama-rag-project\notebooks


In [ ]:
with open(
	"../data/documents/student_handbook.txt",
	"r",
	encoding="utf-8"
) as file:
	text = file.read()

print(text)

Students must register for their courses before the beginning of each semester.

Students are required to attend their scheduled examinations.

Students must complete all graduation requirements before receiving their degree.



----------------

## **11. Text Chunking**

After loading our document into Python, the next step in our RAG pipeline is **`text chunking`**.

Chunking is the process of splitting a large document into smaller pieces of text called **`chunks`**.

### **Why Do We Need Chunking?**

Real-world documents can contain a large amount of information. Sending an entire document to the language model for every question would be inefficient and could include a lot of irrelevant information.

Instead, we want to divide the document into smaller sections so that we can later retrieve only the parts that are relevant to a user's question.

For example:

```text
Document
│
├── Course Registration
├── Examination Rules
├── Attendance Requirements
├── Graduation Requirements
└── Student Services
```

If the user asks:

> "What are the graduation requirements?"

we want the retrieval system to find the section related to **Graduation Requirements** rather than processing the entire document.

### **Our Goal**

The goal of this step is to transform:

```text
Large Document
	  ↓
Smaller Chunks
```

These chunks will become the basic units that we later convert into **embeddings** and store in a **vector database**.

### **RAG Pipeline Progress**

At this point, our pipeline looks like:

```text
Documents
	↓
Document Loading
	↓
Chunking          ← We are here
	↓
Embeddings
	↓
Vector Database
	↓
Retrieve Relevant Information
	↓
Context + User Question
	↓
Ollama / Qwen 2.5 3B
	↓
Answer
```

In the next step, we will implement a simple chunking method and examine how the original document is divided into smaller pieces.


In [13]:
len(text)

227

In [ ]:
chunk_size = 100

chunks = []

for i in range(0, len(text), chunk_size):
	#print(f"i:{i}, i + chunk size: {i + chunk_size}")
	chunk = text[i:i + chunk_size]
	chunks.append(chunk)

chunks

['\ufeffStudents must register for their courses before the beginning of each semester.\n\nStudents are requi',
 'red to attend their scheduled examinations.\n\nStudents must complete all graduation requirements befo',
 're receiving their degree.\n']

As you can see the output is not that much clear so let's make it more `readable`.

In [ ]:
for i, chunk in enumerate(chunks):
	print(f"Chunk {i + 1}:")
	print(chunk)
	print("-" * 50)

Chunk 1:
Students must register for their courses before the beginning of each semester.

Students are requi
--------------------------------------------------
Chunk 2:
red to attend their scheduled examinations.

Students must complete all graduation requirements befo
--------------------------------------------------
Chunk 3:
re receiving their degree.

--------------------------------------------------


### **Chunk Overlap**

Simple character-based chunking can sometimes split words, sentences, or ideas across different chunks.

For example:

```text
Chunk 1:
Students must complete all graduation require

Chunk 2:
ments before receiving their degree.
```

This can make individual chunks less meaningful because important information may be divided between them.

To reduce this problem, we can use **chunk overlap**.

Chunk overlap means that a small portion of one chunk is repeated in the next chunk.

For example:

```text
Chunk 1:
Students must complete all graduation requirements
before receiving their degree.

Chunk 2:
before receiving their degree.
Students must also...
```

The repeated text provides additional context between neighboring chunks.

### **Why Is Overlap Useful?**

Overlap helps preserve context when information spans across chunk boundaries.

This becomes particularly useful during retrieval because the retrieved chunk is more likely to contain enough surrounding context to represent the relevant information.

### **Concept**

```text
Chunk 1
┌──────────────────────────────┐
│        Main Content          │
└──────────────┬───────────────┘
			   │
			   │ Overlap
			   ▼
┌──────────────────────────────┐
│ Overlapping │  New Content   │
└──────────────────────────────┘
Chunk 2
```



In the next step, we will implement chunk overlap and compare it with our previous chunking approach.


In [ ]:
def split_text(text, chunk_size = 100, chunk_overlap = 20):

	chunks = []

	start = 0
 
	while start < len(text):  # 0 < 277 --> True || 80 < 277 --> True || 240 < 277 --> True || 320 < 277 --> False --> exit 
 
		end = start + chunk_size  # end = 0 + 100 --> end = 100 || end = 80 + 100 = 180   || end = 160 + 100 = 260   || end = 240 + 100 = 340
		chunk = text[start : end] # chunk = text[0:100]         || chunk = text[80 : 180] || chunk = text[160 : 260] || chunk = text[240 : 340]
		chunks.append(chunk)

		start = end - chunk_overlap  # start = 100 - 20 = 80    || start = 180 - 20 = 160 || start = 260 - 20 = 240  || start = 340 - 20 = 320

	return chunks 

In [ ]:
chunks = split_text(
	text,
	chunk_size=100,
	chunk_overlap=20
)

In [ ]:
for i, chunk in enumerate(chunks):
	print(f"chunk {i+1} :")
	print(chunk)
	print('-'*50)


chunk 1 :
Students must register for their courses before the beginning of each semester.

Students are requi
--------------------------------------------------
chunk 2 :


Students are required to attend their scheduled examinations.

Students must complete all graduati
--------------------------------------------------
chunk 3 :
omplete all graduation requirements before receiving their degree.

--------------------------------------------------


-------------------------

## **12. Text Embeddings**

After splitting our document into smaller chunks, the next step in our RAG pipeline is to convert these text chunks into **embeddings**.

An embedding is a numerical representation of text that captures information about its semantic meaning.

For example:

```text
"What are the graduation requirements?"
				↓
		Embedding Model
				↓
	 [0.12, -0.45, 0.88, ...]
```

The resulting vector contains numerical values that represent the text in a high-dimensional vector space.

### **Why Do We Need Embeddings?**

RAG systems need a way to determine which document chunks are most relevant to a user's question.

We can convert both the document chunks and the user's question into embeddings and then compare their vectors.

For example:

```text
Document Chunks
	  ↓
Embedding Model
	  ↓
Chunk Vectors

User Question
	  ↓
Embedding Model
	  ↓
Question Vector

Question Vector
	  ↓
Similarity Comparison
	  ↓
Most Relevant Chunks
```

If two pieces of text have similar meanings, their embedding vectors are generally closer together in the vector space.

This allows us to perform **semantic search**, where we search for meaning rather than relying only on exact keyword matches.

### **Embedding vs. Language Model**

Our project uses two different types of models for different purposes:

* **Embedding Model:** Converts text into numerical vectors for similarity search.
* **Qwen 2.5 3B:** Generates the final answer using the retrieved information.

The overall process will eventually look like:

```text
Documents
	↓
Chunking
	↓
Embeddings
	↓
Vector Database
	↓
Retrieve Relevant Information
	↓
Context + User Question
	↓
Qwen 2.5 3B
	↓
Answer
```

### **Objective**

In this section, we will generate embeddings for our document chunks and inspect the resulting vectors.

These embeddings will later be used to build our semantic retrieval system.


### **Installing the Embedding Model**

For this project, we use **`nomic-embed-text`** as our local embedding model.

Unlike `qwen2.5:3b`, which is used to generate responses, `nomic-embed-text` is responsible for converting text into numerical vectors that can be used for semantic search.

The model must be downloaded locally through Ollama before generating embeddings.

If you have not downloaded the model yet, run the following command in your terminal:

```powershell
ollama pull nomic-embed-text
```

After the download is complete, you can verify that the model is available by running:

```powershell
ollama list
```

You should see `nomic-embed-text` in the list of installed models.

### **Why Do We Need This Model?**

Our RAG pipeline uses two different models for two different tasks:

* **`nomic-embed-text`** → converts text into embeddings for semantic search.
* **`qwen2.5:3b`** → generates the final answer using the retrieved context.

This separation allows the retrieval and generation stages of the RAG pipeline to use models designed for their specific tasks.


In [8]:
import ollama

In [ ]:
embedding_response  = ollama.embed(
	model= "nomic-embed-text",
	input= chunks[0]
)

In [21]:
embedding_response

EmbedResponse(model='nomic-embed-text', created_at=None, done=None, done_reason=None, total_duration=841098600, load_duration=789373400, prompt_eval_count=19, prompt_eval_duration=None, eval_count=None, eval_duration=None, embeddings=[[-0.0317717, 0.008476011, -0.18149692, 0.028415533, 0.04869547, 0.052696344, 0.03629947, 0.025471352, 0.022550317, -0.024567705, -0.0146597, 0.039017476, 0.06753937, 0.0178144, -0.026160197, -0.036287356, 0.004648722, -0.02413485, -0.0836744, -0.015785186, -0.0393833, -0.04764778, -0.032317095, -0.009213827, 0.058126245, 0.056566134, -0.026166055, -0.047708534, -0.013863461, 0.032325145, -0.00075172354, -0.023700666, 0.059595525, 0.00079665263, 0.001738852, -0.07784064, -0.007434742, -0.00086662144, -0.046672445, -0.046900444, 0.08710543, -0.01864541, 0.034378327, -0.020260006, -0.013054639, 0.027306, 0.061361983, -0.008444095, 0.06040033, 0.0017531557, -0.017250743, -0.056325443, -0.055979434, -0.010274218, 0.011993739, 0.037167724, -0.0061187707, 0.0317

In [19]:
# Take only embeddings
#embedding_response.embeddings

In [50]:
# Take only the Victor 
embedding = embedding_response.embeddings[0]

#embedding

In [51]:
len(embedding)

768

In [25]:
# Print only the first 10 
embedding[:10]

[-0.0317717,
 0.008476011,
 -0.18149692,
 0.028415533,
 0.04869547,
 0.052696344,
 0.03629947,
 0.025471352,
 0.022550317,
 -0.024567705]

In [26]:
# Put them together 
embedding = embedding_response.embeddings[0]

print("Embedding dimensions:", len(embedding))
print("First 10 values:", embedding[:10])

Embedding dimensions: 768
First 10 values: [-0.0317717, 0.008476011, -0.18149692, 0.028415533, 0.04869547, 0.052696344, 0.03629947, 0.025471352, 0.022550317, -0.024567705]


In [ ]:
def embed_chunks(chunks):

	embeddings = []

	for chunk in chunks:
		response  = ollama.embed(
			model= "nomic-embed-text",
			input= chunk
		)
		embedding = response.embeddings[0]
		embeddings.append(embedding)
		
	return embeddings

In [21]:
embeddings = embed_chunks(chunks)

In [22]:
print("Number of chunks:", len(chunks))
print("Number of embeddings:", len(embeddings))
print("Dimensions of first embedding:", len(embeddings[0]))

Number of chunks: 3
Number of embeddings: 3
Dimensions of first embedding: 768


**Note:** This function converts each text chunk into a 768-dimensional embedding vector using `nomic-embed-text`. We extract the vector using `response.embeddings[0]` and store it in a list, keeping the embeddings in the same order as their original chunks.


------------

## **13. Cosine Similarity**

After generating embeddings for our document chunks, we need a way to determine which chunks are most relevant to a user's question.

We can do this by comparing the embedding of the user's question with the embeddings of our document chunks.

### **The Problem**

Suppose we have:

```text
Question
   ↓
Question Embedding

Chunk 1
   ↓
Chunk 1 Embedding

Chunk 2
   ↓
Chunk 2 Embedding

Chunk 3
   ↓
Chunk 3 Embedding
```

We need to determine which chunk is most similar to the question.

### **Cosine Similarity**

**Cosine similarity** measures the similarity between two vectors based on the angle between them.

If two vectors point in a similar direction, their cosine similarity is high.

Conceptually:

```text
Similar direction
	↓
High similarity
```

For example:

```text
Question → Chunk 1 = 0.21
Question → Chunk 2 = 0.87
Question → Chunk 3 = 0.34
```

In this case, **Chunk 2** is the most relevant because it has the highest similarity score.

Cosine similarity is commonly represented using values between `-1` and `1`:

* `1` → vectors point in the same direction.
* `0` → vectors are orthogonal.
* `-1` → vectors point in opposite directions.

### **Why Do We Need It?**

Our goal in RAG is to retrieve the document chunks that are most relevant to the user's question.

The process will be:

```text
User Question
	↓
Question Embedding
	↓
Compare with Chunk Embeddings
	↓
Cosine Similarity
	↓
Rank Chunks by Similarity
	↓
Retrieve the Most Relevant Chunks
```

This is the foundation of **semantic search** in our RAG system.

### **Objective**

In this section, we will calculate cosine similarity between our question embedding and the embeddings of our document chunks.


### **Cosine Similarity Formula**

The cosine similarity between two vectors `A` and `B` is calculated using:

$$
\text{similarity}(A,B)
=
\frac{A \cdot B}{\|A\| \|B\|}
$$

Where:

* $A \cdot B$ → the **dot product** of the two vectors.
* $|A|$ → the **magnitude** (length) of vector `A`.
* $|B|$ → the **magnitude** (length) of vector `B`.


In [31]:
import sys

print(sys.executable)

c:\Users\User\ollama-rag-project\.venv\Scripts\python.exe


In [33]:
import torch
import numpy as np

print(torch.__version__)
print(np.__version__)

2.14.1+cpu
2.4.6


In [ ]:
def cosine_similarity(a, b):

	a = torch.tensor(a)
	b = torch.tensor(b)

	dot = a @ b
	magnitude_a = torch.linalg.norm(a)
	magnitude_b = torch.linalg.norm(b)

	result = dot / (magnitude_a * magnitude_b)

	return result

In [35]:
a = torch.tensor([1.0, 0.0])
b = torch.tensor([1.0, 0.0])

print(cosine_similarity(a, b))

tensor(1.)


C:\Users\User\AppData\Local\Temp\ipykernel_17064\4119807734.py:3: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.detach().clone() or sourceTensor.detach().clone().requires_grad_(True), rather than torch.tensor(sourceTensor).
  a = torch.tensor(a)
C:\Users\User\AppData\Local\Temp\ipykernel_17064\4119807734.py:4: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.detach().clone() or sourceTensor.detach().clone().requires_grad_(True), rather than torch.tensor(sourceTensor).
  b = torch.tensor(b)


In [36]:
a = torch.tensor([1.0, 0.0])
b = torch.tensor([0.0, 1.0])

print(cosine_similarity(a, b))

tensor(0.)


C:\Users\User\AppData\Local\Temp\ipykernel_16280\4119807734.py:3: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.detach().clone() or sourceTensor.detach().clone().requires_grad_(True), rather than torch.tensor(sourceTensor).
  a = torch.tensor(a)
C:\Users\User\AppData\Local\Temp\ipykernel_16280\4119807734.py:4: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.detach().clone() or sourceTensor.detach().clone().requires_grad_(True), rather than torch.tensor(sourceTensor).
  b = torch.tensor(b)


-----------

## **14. From Similarity to Retrieval**

We can now calculate the **cosine similarity** between two embedding vectors.

However, calculating the similarity between two vectors is not enough for our RAG system. We need to use it to determine **which document chunks are relevant to a user's question**.

### **The Next Step**

First, we convert the user's question into an embedding using `nomic-embed-text`.

Then, we compare the question embedding with the embedding of every document chunk.

```text
User Question
	  ↓
Question Embedding
	  ↓
┌──────────────────────────────┐
│ Compare with Chunk Embeddings│
│                              │
│ Question ↔ Chunk 1 → Score   │
│ Question ↔ Chunk 2 → Score   │
│ Question ↔ Chunk 3 → Score   │
└──────────────────────────────┘
	  ↓
Similarity Scores
	  ↓
Rank the Chunks
	  ↓
Retrieve the Most Relevant Chunk(s)
```

For example:

```text
Question → "When do students register for courses?"

Chunk 1 → 0.82
Chunk 2 → 0.15
Chunk 3 → 0.31
```

Since **Chunk 1** has the highest similarity score, it is considered the most relevant chunk for the question.

### **Retrieval**

This process is called **retrieval**.

Our goal is to create a function that:

1. Takes a user's question.
2. Converts the question into an embedding.
3. Compares the question embedding with all document chunk embeddings.
4. Calculates a similarity score for each chunk.
5. Identifies the most relevant chunks.
6. Returns the relevant chunks so they can later be provided to the LLM as context.

### **Updated RAG Pipeline**

We have now completed:

```text
Documents
	↓
Document Loading
	↓
Chunking
	↓
Embeddings
	↓
Cosine Similarity
```

Next, we will implement:

```text
Similarity Scores
	↓
Retrieval
	↓
Relevant Context
	↓
LLM
	↓
Final Answer
```

This is the point where our embedding and similarity work starts becoming an actual **RAG retrieval system**.


In [ ]:
def retrieve_relevant_chunks(question, chunks, embeddings):

	response = ollama.embed(
		model="nomic-embed-text",
		input=question
	)

	question_embedding = response.embeddings[0]

	similarities = []

	for embedding in embeddings:
		similarity = cosine_similarity(question_embedding, embedding)

		similarities.append(similarity)

	best_index = similarities.index(max(similarities))

	# Get the chunk of the best index
	best_chunk = chunks[best_index]

	return best_chunk

In [ ]:
question = "When do students register for courses?"

result = retrieve_relevant_chunks(
	question,
	chunks,
	embeddings
)

print(result)

Students must register for their courses before the beginning of each semester.

Students are requi


> **Note:** The retrieval step is working correctly. However, our current chunking strategy is still very simple and can cut words or sentences because it is based on a fixed character size. We will improve the chunking strategy later to produce cleaner and more meaningful chunks.


-----------------------

## **15. Generation with Context**

We have now implemented the **retrieval** part of our RAG pipeline.

Given a user's question, our system can:

1. Convert the question into an embedding.
2. Compare it with the embeddings of our document chunks.
3. Find the most relevant chunk.
4. Return that chunk as the retrieved context.

For example:

```text
Question:
"When do students register for courses?"

		  ↓

Retrieved Context:
"Students must register for their courses before the
beginning of each semester."
```

However, retrieving the relevant information is not enough. We still need an **LLM** to use this information and generate a natural-language answer.

### **Why Do We Need Context?**

If we send the question directly to the LLM:

```text
Question
	↓
Qwen 2.5
	↓
Answer
```

the model will mainly rely on the knowledge it learned during training.

In a RAG system, instead, we first retrieve relevant information from our own documents and provide it to the LLM as **context**.

```text
User Question
		↓
	Retrieval
		↓
Relevant Context
		+
User Question
		↓
	Qwen 2.5
		↓
  Final Answer
```

This allows the model to generate an answer based on the information retrieved from our documents.

### **Retrieval-Augmented Generation**

This is where the three parts of **RAG** come together:

* **Retrieval** → Find relevant information from our documents.
* **Augmented** → Add the retrieved information to the user's question as context.
* **Generation** → Use the LLM to generate the final answer.

```text
Documents
	 ↓
  Chunking
	 ↓
 Embeddings
	 ↓
  Retrieval
	 ↓
Relevant Context
	 +
User Question
	 ↓
	Qwen 2.5
	 ↓
 Final Answer
```

### **Context + Question**

Before sending the request to the LLM, we will construct a prompt containing both the retrieved context and the user's question.

Conceptually, it will look like this:

```text
Context:
Students must register for their courses before the beginning
of each semester.

Question:
When do students register for courses?
```

The LLM can then use the provided context to generate the answer:

```text
Students must register for their courses before the beginning
of each semester.
```

### **Objective**

In this section, we will:

1. Retrieve the relevant context for a user's question.
2. Combine the context with the question.
3. Send the combined prompt to `qwen2.5:3b`.
4. Generate the final answer using the retrieved information.

This will complete the basic end-to-end RAG pipeline.


In [38]:
prompt = f"""
Context:
{result}

Question:
{question}
"""

In [39]:
print(prompt)


Context:
Students must register for their courses before the beginning of each semester.

Students are requi

Question:
When do students register for courses?



In [ ]:
response = ollama.chat(
	model="qwen2.5:3b",
	messages=[
		{
			"role": "user",
			"content": prompt
		}
	]
)

print(response["message"]["content"])

Students must register for their courses before the beginning of each semester.


--------------

## **16. How Context Influences Generation**

So far, we have implemented the **retrieval** part of our RAG pipeline.

Our system can now take a user's question and retrieve the document chunk that is most relevant to that question.

However, retrieving information is only half of the RAG process.

We now need to pass the retrieved information to a language model so that it can generate a natural-language answer.

The basic flow is:

```text
User Question
	↓
Retrieve Relevant Chunk
	↓
Retrieved Context
	↓
Context + Question
	↓
Qwen 2.5 3B
	↓
Generated Answer
```

In this section, we will investigate how providing context to the language model influences its generated answer.

We will also test what happens when the answer **exists** in the retrieved context and when it **does not exist**.


### **Why Do We Need Context?**

A language model such as `qwen2.5:3b` already contains a large amount of knowledge from its training data.

However, the goal of our RAG system is not simply to ask the model a question and let it answer from its general knowledge.

Instead, we want the model to use information retrieved from **our own documents**.

For example, our document contains:

```text
Students must register for their courses before the beginning of each semester.

Students are required to attend their scheduled examinations.

Students must complete all graduation requirements before receiving their degree.
```

If the user asks:

```text
When do students register for courses?
```

the retrieval system should find the relevant part of the document.

We can then provide that retrieved text to the language model as **context**.

The model can use this context to generate the final answer.

Therefore, the role of each component is:

* **Retriever** → finds relevant information.
* **Context** → provides that information to the LLM.
* **LLM** → generates a natural-language answer.


### **Context + Question**

The LLM needs two important pieces of information:

1. **Context** — the relevant information retrieved from our documents.
2. **Question** — the question asked by the user.

We combine both into a prompt.

Conceptually:

```text
Context:
[Retrieved document information]

Question:
[User question]

Answer:
```

We also give the model an instruction telling it how to use the context.

For example:

```text
Answer the question using only the provided context.

If the answer is not contained in the context, say:
"I don't know."
```

This is important because we do not want the model to invent information that is not present in our documents.


In [ ]:
# Test the Retrieval
question = "When do students register for courses?"

context = retrieve_relevant_chunks(
	question,
	chunks,
	embeddings
)

print("Retrieved Context:")
print(context)

Retrieved Context:
Students must register for their courses before the beginning of each semester.

Students are requi


### **Building the Prompt**

Now that we have the retrieved context, we need to combine it with the user's question.

We will create a prompt using an f-string so that the retrieved context and question are inserted dynamically.

The prompt will contain three main parts:

```text
Instruction
	↓
Context
	↓
Question
```

The instruction tells the model to use only the provided context.

The context contains the information retrieved from our documents.

The question tells the model what information the user is asking for.


In [45]:
# Build the Prompt

prompt = f"""
Answer the question using only the provided context.

If the answer is not contained in the context, say:
"I don't know."

Context:
{context}

Question:
{question}

Answer:
"""

print(prompt)


Answer the question using only the provided context.

If the answer is not contained in the context, say:
"I don't know."

Context:
Students must register for their courses before the beginning of each semester.

Students are requi

Question:
When do students register for courses?

Answer:



### **Generating the Answer**

We can now send the prompt to `qwen2.5:3b`.

We use `ollama.chat()` because we want the model to generate a response based on the prompt.

The model receives the retrieved context and the user's question together.


In [ ]:
# Generate Answer
response = ollama.chat(
	model="qwen2.5:3b",
	messages=[
		{
			"role": "user",
			"content": prompt
		}
	]
)

answer = response["message"]["content"]

print("Answer:")
print(answer)

Answer:
Students register for courses before the beginning of each semester.


### **Understanding the Generation Flow**

At this point, our system has connected retrieval with generation.

The complete process is:

```text
User Question
	↓
retrieve_relevant_chunks()
	↓
Relevant Context
	↓
Create Prompt
	↓
ollama.chat()
	↓
Qwen 2.5 3B
	↓
Generated Answer
```

The important part is that the LLM is no longer receiving only the user's question.

It receives:

```text
Context + Question
```

This is what allows the model to generate an answer grounded in our documents.


### **Test 1 — Answer Exists in the Context**

First, we will test a question whose answer is clearly present in our document.

```text
When do students register for courses?
```

Our document contains:

```text
Students must register for their courses before the beginning of each semester.
```

Therefore, we expect the model to generate an answer based on this information.


In [ ]:
question = "When do students register for courses?"

context = retrieve_relevant_chunks(
	question,
	chunks,
	embeddings
)

prompt = f"""
Answer the question using only the provided context.

If the answer is not contained in the context, say:
"I don't know."

Context:
{context}

Question:
{question}

Answer:
"""

response = ollama.chat(
	model="qwen2.5:3b",
	messages=[
		{
			"role": "user",
			"content": prompt
		}
	]
)

answer = response["message"]["content"]

print("Retrieved Context:")
print(context)

print("\nAnswer:")
print(answer)

Retrieved Context:
Students must register for their courses before the beginning of each semester.

Students are requi

Answer:
Students register for courses before the beginning of each semester.


The answer was present in the retrieved context, and the model generated an answer using that information.

This demonstrates the basic **retrieval → context → generation** workflow.

```text
Question
	↓
Retrieve Relevant Chunk
	↓
Context contains the answer
	↓
Qwen
	↓
Answer based on context
```

The retrieval stage provides the relevant information, while the generation stage turns that information into a natural-language response.


### **Test 2 — Answer Does Not Exist in the Context**

Now we will test the opposite situation.

We will ask a question that has nothing to do with our document:

```text
What is the capital of France?
```

Our document contains information about:

* Course registration
* Examinations
* Graduation requirements

It does not contain information about France.

Therefore, the answer should not be generated from our document.

We instructed the model:

```text
If the answer is not contained in the context, say:
"I don't know."
```

Let's test this behavior.


In [ ]:
question = "What is the capital of France?"

context = retrieve_relevant_chunks(
				question,
				chunks,
				embeddings
)

prompt = f"""
Answer the question using only the provided context.

If the answer is not contained in the context, say:
"I don't know."

Context:
{context}

Question:
{question}

Answer:
"""

response = ollama.chat(
				model="qwen2.5:3b",
				messages=[
								{
												"role": "user",
												"content": prompt
								}
				]
)

answer = response["message"]["content"]

print("Retrieved Context:")
print(context)

print("\nAnswer:")
print(answer)

Retrieved Context:
Students must register for their courses before the beginning of each semester.

Students are requi

Answer:
I don't know.


The retrieved context did not contain information about the capital of France.

The model therefore returned:

```text
I don't know.
```

This is the behavior we want in a basic RAG system.

Instead of answering from its general knowledge, the model followed our instruction to use only the provided context.


### **What Did We Learn?**

These two experiments demonstrate how context influences the generation of an LLM.

When the answer exists in the context:

```text
Question
	↓
Retrieved Context
	↓
Context contains answer
	↓
LLM
	↓
Answer
```

When the answer does not exist in the context:

```text
Question
	↓
Retrieved Context
	↓
Context does not contain answer
	↓
LLM
	↓
"I don't know."
```

This is an important concept in **Retrieval-Augmented Generation**.

The LLM is not expected to answer the question using only its pretrained knowledge.

Instead, the RAG system first retrieves relevant information and provides that information to the LLM as context.

The LLM then uses that context to generate the final response.


-----

### **Reducing Hallucinations**

One of the goals of using retrieved context is to reduce **hallucinations**.

A hallucination occurs when a language model generates information that is incorrect or unsupported by the provided information.

For example, if our document does not contain information about a university's refund policy, we do not want the model to invent a refund policy.

Our prompt therefore tells the model:

```text
Answer the question using only the provided context.

If the answer is not contained in the context, say:
"I don't know."
```

This can help reduce hallucinations.

However, prompt instructions alone do not guarantee that a model will always behave perfectly.

The quality of the final answer also depends on the quality of:

* Document chunking
* Embeddings
* Retrieval
* Retrieved context
* Prompt design
* Language model


### **Important Limitation**

Our current experiment works with a very small document and a simple retrieval strategy.

There are several limitations:

* We currently retrieve only **one chunk**.
* Our chunking strategy is based on a fixed character size.
* Some chunks can cut words or sentences.
* We do not currently use a similarity threshold.
* We do not rank multiple retrieved chunks.
* The LLM may still hallucinate in more difficult situations.

These are not problems we need to solve immediately.

The purpose of this section is to understand the basic relationship between **retrieval, context, and generation**.

We will improve the retrieval and chunking strategies later.


### **Current RAG Pipeline**

At this stage, our RAG pipeline looks like this:

```text
Documents
	↓
Document Loading
	↓
Chunking
	↓
Embeddings
	↓
Cosine Similarity
	↓
Retrieval
	↓
Relevant Context
	↓
Context + User Question
	↓
Qwen 2.5 3B
	↓
Generated Answer
```

We have now connected the **retrieval** and **generation** stages.

The core idea of our RAG system is:

> **Retrieve relevant information first, then give that information to the LLM as context for generating the answer.**

This is the foundation of our local RAG application.


---

## **17. End-to-End RAG Pipeline**

So far, we have built the main components of our RAG system separately.

We can now:

* Load documents.
* Split documents into chunks.
* Generate embeddings.
* Calculate cosine similarity.
* Retrieve relevant chunks.
* Provide retrieved context to the language model.
* Generate an answer using `qwen2.5:3b`.

However, these components are still being used separately.

For example, we currently need to manually retrieve the relevant context and then build a prompt before sending it to the language model.

Our next goal is to connect these components into a reusable **end-to-end RAG pipeline**.

### **What Are We Building?**

The goal is to make the RAG system handle the complete process automatically:

```text
User Question
	↓
Retrieve Relevant Context
	↓
Build Prompt
	↓
Qwen 2.5 3B
	↓
Generated Answer
```

Instead of manually performing each step, we want to create reusable functions that handle these operations for us.

### **Separating Retrieval and Generation**

Before creating the complete RAG pipeline, we will first create a function responsible only for the **generation** stage.

This function will receive:

* The user's question.
* The retrieved context.

It will then create the prompt, send it to `qwen2.5:3b`, and return the generated answer.

Conceptually:

```text
Question + Context
	  ↓
   generate_answer()
	  ↓
	Answer
```

This separation is useful because it keeps the different stages of our RAG system independent.

Later, we can combine:

```text
Retrieval
    ↓
Generation
```

into a single end-to-end pipeline.

### **Why Use Functions?**

Breaking the RAG system into smaller functions makes the project easier to:

* `Understand`.
* `Test`.
* `Debug`.
* `Reuse`.
* `Improve later`.

For example, if we want to improve our retrieval strategy later, we can modify the retrieval function without changing the generation logic.

Similarly, we can change the language model or prompt without rewriting the retrieval system.

### **Our Goal**

By the end of this section, we want to move from manually running individual RAG steps to a system where the process can eventually look like:

```python
answer = rag(question)
```

Internally, the system will handle:

```text
User Question
	↓
Question Embedding
	↓
Similarity Search
	↓
Relevant Context
	↓
Prompt Construction
	↓
Qwen 2.5 3B
	↓
Final Answer
```

This will be the foundation for building a more complete RAG application later.


In [ ]:
def generate_answer(question, context):

				prompt = f"""
Answer the question using only the provided context.

If the answer is not contained in the context, say:
"I don't know."

Context:
{context}

Question:
{question}

Answer:
"""

				response = ollama.chat(
								model="qwen2.5:3b",
								messages=[
												{
																"role": "user",
																"content": prompt
												}
								]
				)

				answer = response["message"]["content"]

				return answer

In [ ]:
def rag(question):

				context = retrieve_relevant_chunks(question, chunks, embeddings)

				prompt = f"""
													Answer the question using only the provided context.
													If the answer is not contained in the context, say:
													"I don't know."
													Context:
													{context}

													Question:
													{question}

													Answer:
													"""
				response = ollama.chat(
					model= "qwen2.5:3b",
					messages=[
						{
							"role":"user",
							"content":prompt
						}
					]
				)

				answer = response["message"]["content"]

				return answer

In [46]:
# Test 1 
question = "When do students register for courses?"
rag(question)

'Students must register for their courses before the beginning of each semester.'

In [54]:
# Test 2
question = "What is the capital of France?"
rag(question)

"I don't know."

In [55]:
# Test 3
question = "What are the requirements for receiving a degree?"
rag(question)

'The requirements for receiving a degree are to complete all graduation requirements.'

> **Note:** We have now connected retrieval and generation into a single end-to-end RAG pipeline. The system can take a user question, retrieve the most relevant document chunk, provide it as context to `qwen2.5:3b`, and generate a grounded answer.
>
> Our current retrieval strategy returns only one chunk, which is a limitation we will improve in the next section.


--------------------

## **18. Improving Retrieval**

Our first retrieval implementation works, but it has an important limitation: it returns only **one document chunk**.

For small documents this may be enough, but for larger documents, the information needed to answer a question may be spread across multiple chunks.

### **The Limitation of Single-Chunk Retrieval**

Our current retrieval process looks like this:

```text
User Question
	↓
Question Embedding
	↓
Compare with all chunk embeddings
	↓
Find highest similarity
	↓
Return one chunk
```

The problem is that the most relevant information may not exist entirely inside a single chunk.

For example:

```text
Chunk 22 → contains the beginning of the information
Chunk 23 → contains the main information
Chunk 24 → contains additional details
```

If we retrieve only `Chunk 23`, the language model may not have enough context to produce the best answer.

### **Top-K Retrieval**

To improve this, we can retrieve the **top K most relevant chunks** instead of retrieving only the single most relevant chunk.

For example, if we use:

```text
top_k = 3
```

the retrieval system will return the three chunks with the highest similarity scores.

The new process becomes:

```text
User Question
	↓
Question Embedding
	↓
Compare with all chunk embeddings
	↓
Calculate similarity scores
	↓
Rank chunks by similarity
	↓
Select Top-K chunks
	↓
Relevant Context
```

### **Example**

Suppose our similarity scores are:

```text
Chunk 1 → 0.72
Chunk 2 → 0.91
Chunk 3 → 0.68
Chunk 4 → 0.95
Chunk 5 → 0.84
```

If `top_k = 3`, we want to retrieve:

```text
Chunk 4 → 0.95
Chunk 2 → 0.91
Chunk 5 → 0.84
```

These chunks can then be combined into a single context and passed to the language model.

### **Why Top-K Retrieval Matters**

Retrieving multiple chunks can help the language model access more relevant information.

It can be especially useful when:

* The answer is spread across multiple chunks.
* The document is large.
* Important details are located in different parts of the document.
* The best individual chunk does not contain enough context.

However, retrieving too many chunks is not always better.

If we provide too much irrelevant context, the language model may have more information to process and may become less focused.

Therefore, `top_k` is an important parameter that we can experiment with.

### **Updating Our Retrieval Function**

Our current function is designed to return one chunk:

```python
retrieve_relevant_chunks(question, chunks, embeddings)
```

We will update it so that it accepts a `top_k` parameter:

```python
retrieve_relevant_chunks(
    question,
    chunks,
    embeddings,
    top_k=3
)
```

The function will then:

1. Create an embedding for the user's question.
2. Calculate the similarity between the question and every document chunk.
3. Keep track of the similarity scores and their corresponding chunk indexes.
4. Find the `top_k` most similar chunks.
5. Return those chunks as the retrieved context.

### **Goal of This Section**

By the end of this section, our retrieval system should be able to return multiple relevant chunks instead of only one.

This will make our RAG pipeline more robust and prepare it for working with larger documents.

Our improved pipeline will look like:

```text
Documents
    ↓
Chunking
    ↓
Embeddings
    ↓
Question
    ↓
Similarity Search
    ↓
Top-K Relevant Chunks
    ↓
Context
    ↓
Qwen 2.5 3B
    ↓
Answer
```

> **Note:** `top_k` is a retrieval parameter, not a fixed rule. The best value depends on the document size, chunking strategy, and the type of questions users ask.


In [ ]:
def retrieve_relevant_chunks(question, chunks, embeddings, top_k=3):

				response = ollama.embed(
								model="nomic-embed-text",
								input=question
				)

				question_embedding = response.embeddings[0]

				similarities = []

				for embedding in embeddings:
								similarity = cosine_similarity(question_embedding, embedding)

								similarities.append(similarity)
								
				indexed_similarities = []
								
				for index, value in enumerate(similarities):
								
								indexed_similarities.append((value, index))

				indexed_similarities.sort(reverse=True) # highest to lowestd

				best_indexed = indexed_similarities[0:top_k]  # Get the top-k most relevant chunks

				best_chunks = []

				for value, index in best_indexed:
					# Get the chunk of the best index
					best_chunks.append(chunks[index])

				return best_chunks

In [43]:
result = retrieve_relevant_chunks(
    "When do students register for courses?",
    chunks,
    embeddings
)

print(result)

['\ufeffStudents must register for their courses before the beginning of each semester.\n\nStudents are requi', '\n\nStudents are required to attend their scheduled examinations.\n\nStudents must complete all graduati', 'omplete all graduation requirements before receiving their degree.\n']


In [45]:
context = "\n\n".join(result)
context

'\ufeffStudents must register for their courses before the beginning of each semester.\n\nStudents are requi\n\n\n\nStudents are required to attend their scheduled examinations.\n\nStudents must complete all graduati\n\nomplete all graduation requirements before receiving their degree.\n'

In [50]:
def rag(question):

				context = retrieve_relevant_chunks(question, chunks, embeddings,top_k=3)

				context = "\n\n".join(context)

				prompt = f"""
													Answer the question using only the provided context.
													If the answer is not contained in the context, say:
													"I don't know."
													Context:
													{context}

													Question:
													{question}

													Answer:
													"""
				response = ollama.chat(
					model= "qwen2.5:3b",
					messages=[
						{
							"role":"user",
							"content":prompt
						}
					]
				)

				answer = response["message"]["content"]

				return answer

In [53]:
question = "When do students register for courses?"

answer = rag(question)

print(answer)

Students must register for their courses before the beginning of each semester.


In [54]:
question = "What is the capital of France?"

answer = rag(question)

print(answer)

I don't know.


-----------------

## **19. Better Text Chunking**

Our current chunking strategy splits documents based on a fixed number of characters.

For example, we currently use:

```python
chunk_size = 100
chunk_overlap = 20
```

This approach is simple and useful for learning the fundamentals of RAG. However, it has an important limitation: it does not understand the structure or meaning of the text.

### **The Problem with Character-Based Chunking**

Suppose our document contains:

```text
Students must complete all graduation requirements before receiving their degree.
```

A character-based chunking strategy could split this sentence in the middle:

```text
Chunk 1:
Students must complete all graduation requi

Chunk 2:
rements before receiving their degree.
```

The information is now divided between two chunks.

This can make retrieval less effective because each chunk may contain only part of the original meaning.

### **Why Chunking Matters**

Chunking is one of the most important steps in a RAG pipeline.

The chunks are the units that we:

* Convert into embeddings.
* Compare with the user's question.
* Retrieve from the document collection.
* Provide to the language model as context.

Therefore, the quality of our chunks can directly affect the quality of retrieval and the final answer.

A useful chunk should ideally contain a meaningful piece of information.

### **A Better Strategy**

Instead of immediately splitting the document based on a fixed number of characters, we can first use the natural structure of the document.

For example, if the document contains multiple paragraphs:

```text
Paragraph 1

Paragraph 2

Paragraph 3
```

We can create chunks based on those paragraphs:

```text
Chunk 1 → Paragraph 1
Chunk 2 → Paragraph 2
Chunk 3 → Paragraph 3
```

This preserves more of the original meaning.

### **Paragraph-Based Chunking**

For our current project, we will start with a simple paragraph-based strategy.

The idea is:

```text
Document
    ↓
Split into paragraphs
    ↓
Remove empty paragraphs
    ↓
Use paragraphs as chunks
```

This is intentionally simple.

We are not trying to build a production-grade chunking system yet. The goal is to understand how different chunking strategies affect a RAG pipeline.

### **Comparing the Two Strategies**

Our original approach:

```text
Document
    ↓
Fixed character size
    ↓
Character chunks
    ↓
Embeddings
```

Our new approach:

```text
Document
    ↓
Paragraph boundaries
    ↓
Meaningful chunks
    ↓
Embeddings
```

### **Important Trade-off**

Paragraph-based chunking is not always better.

If a paragraph is extremely large, using the entire paragraph as one chunk may also be a problem.

For example:

```text
Very large paragraph
        ↓
One huge chunk
        ↓
Large embedding/context
```

Later, we can build a more advanced strategy that combines semantic boundaries with a maximum chunk size.

For now, paragraph-based chunking gives us a better foundation while keeping the implementation easy to understand.

### **Our Goal**

By the end of this section, we want to replace our current character-based chunking function with a paragraph-based chunking function.

The updated pipeline will be:

```text
Document
    ↓
Paragraph-Based Chunking
    ↓
Embeddings
    ↓
Top-K Retrieval
    ↓
Relevant Context
    ↓
Qwen 2.5 3B
    ↓
Answer
```

> **Note:** Changing the chunking strategy also means that our existing embeddings are no longer valid for the new chunks. After creating the new chunks, we need to generate new embeddings for them.


In [60]:
def split_text(text):

				chunks = text.split("\n\n")

				new_chunks = []

				for chunk in chunks:
					if  chunk.strip():
						new_chunks.append(chunk)
					
					
				return new_chunks


In [61]:
print(split_text(text=text))

['\ufeffStudents must register for their courses before the beginning of each semester.', 'Students are required to attend their scheduled examinations.', 'Students must complete all graduation requirements before receiving their degree.\n']


In [62]:
# New emdeddings 

embeddings = embed_chunks(chunks)

In [64]:
# check them 
print("Number of chunks:", len(chunks))
print("Number of embeddings:", len(embeddings))
print("Dimensions of first embedding:", len(embeddings[0]))

Number of chunks: 3
Number of embeddings: 3
Dimensions of first embedding: 768
